enn.py

```diff
@@ -33,41 +36,112 @@ def __init__(self, *args, **kwargs):
-            y_pos = np.arange(len(rsacms))
-
-            # Create bars
-            plt.bar(y_pos, loss_value)
-
-            # Create names on the x-axis
-            plt.xticks(y_pos, rsacms)
-            plt.title(f'MSELoss in {module_name}')
-            plt.plot()
-            plt.xlabel('Number of RSACM')
-            plt.ylabel('MSELoss value')
-            plt.savefig(f'./graphs/{module_name}.png')
-            plt.clf()
+            losses_per_layers_student_train = {module_name: [[] for i in range(0,self.args.number_of_rsacm)] for module_name in self.model.replaced_modules}
+            for i, (images, labels) in enumerate(tqdm(train_loader)):
+                    images, labels = images.to(self.args.device), labels.to(self.args.device)
+                    self.teacher_model(images)
+                    loss_distill = 0
+                    for index, module_name in enumerate(self.model.replaced_modules):
+                        get_module_by_name(self.model, module_name)(self.teacher_model.inputs[index][0])
+                        for rsacm_index, rsacm_block_output in enumerate(get_module_by_name(self.model, module_name).x_list):
+                            loss_distill += self.criterion_distilation(rsacm_block_output, self.teacher_model.activation[index])
+                            losses_per_layers_student_train[module_name][rsacm_index].append(self.criterion_distilation(rsacm_block_output, self.teacher_model.activation[index]))
+            # print(losses_per_layers_student)
+            losses_per_layers_student_mean_train = {module_name: [(sum(losses_per_layers_student_train[module_name][i]) / len(losses_per_layers_student_train[module_name][i])).item() for i in range(0, self.args.number_of_rsacm)] for module_name in self.model.replaced_modules}
+            for module_name in losses_per_layers_student_mean_valid:
+                # Make a random dataset:
+                rsacms = range(1, self.args.number_of_rsacm+1)
+                loss_value_valid = losses_per_layers_student_mean_valid[module_name]
+                loss_value_train = losses_per_layers_student_mean_train[module_name]
+                y_pos = np.arange(len(rsacms))
+                width = 0.3       
+
+                # Create bars
+                plt.bar(y_pos, loss_value_valid, width, label='valid')
+                plt.bar(y_pos+width, loss_value_train, width, label='train')
+
+                # Create names on the x-axis
+                plt.legend(loc='best')
+                plt.xticks(y_pos, rsacms)
+                plt.title(f'MSELoss in {module_name}')
+                plt.plot()
+                plt.xlabel('Number of RSACM')
+                plt.ylabel('MSELoss value')
+                plt.savefig(f'./graphs/{module_name}.png')
+                plt.clf()
+
+
+    def train_gating_networks(self, train_loader, valid_loader):
+
+        accuracy_student = eval(valid_loader, self.model, self.args)
```

Comment:
This is acc with all RSACMs enabled, right?

Reply:
Yes.